### PR TITLE
PR: Liquidation of borrow

### PR DESCRIPTION
**Pull request by convowork1**

### CODE REVIEW

File: contracts/dtoken/src/liquidation.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        liquidator: AccountId,
+        liquidation_amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            Contract::custom_fail_log(
+                String::from("withdraw_fail"),
+                env::signer_account_id(),
+                Balance::from(liquidation_amount.0),
+                format!(
+                    "Liquidation failed, while tried to call controller ({:?})",
```
Comment (markblaize) [inline contracts/dtoken/src/liquidation.rs:53]: @convowork1 we cannot use `:?` in production as it inteded for debug only
"Liquidation failed, while tried to call controller ({:?})" -> "failed to liquidate on {}", self.get_controller_address()

File: contracts/dtoken/src/liquidation.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        &mut self,
+        borrower: AccountId,
+        borrowing_dtoken: AccountId,
+        collateral_dtoken: AccountId,
+        liquidator: AccountId,
+        liquidation_amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            Contract::custom_fail_log(
+                String::from("withdraw_fail"),
```
Comment (markblaize) [inline contracts/dtoken/src/liquidation.rs:49]: @convowork1 withdraw_fail -> liquidation_fail

File: contracts/dtoken/src/liquidation.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                env::signer_account_id(),
+                Balance::from(liquidation_amount.0),
+                format!(
+                    "Liquidation failed, while tried to call controller ({:?})",
+                    self.get_controller_address()
+                ),
+            );
+            env::panic_str("");
```
Comment (markblaize) [inline contracts/dtoken/src/liquidation.rs:57]: @convowork1 panic should still have meaningful message, e.g. "failed to liquidate"

File: contracts/dtoken/src/liquidation.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            liquidation_amount,
+            self.get_controller_address(),
+            NO_DEPOSIT,
+            self.terra_gas(20),
+        ))
+        .into()
+    }
+
+    #[private]
+    pub fn liquidation_repay_borrows_callback(
```
Comment (markblaize) [inline contracts/dtoken/src/liquidation.rs:82]: @convowork1 are we using this anywhere?

File: contracts/dtoken/src/ft.rs [hunk 5]
```diff
@@ -31,20 +31,41 @@ impl FungibleTokenReceiver for Contract {
...
         // TODO: In future make action not a single one, but array in JSON message
         let action: &str = msg["action"].as_str().unwrap();
         match action {
             "SUPPLY" => self.supply(amount),
             "REPAY" => self.repay(amount),
+            "LIQUIDATION" => {
+                let memo_data = msg["memo"].clone();
+
+                let borrower =
+                    AccountId::new_unchecked(String::from(memo_data["borrower"].as_str().unwrap()));
```
Comment (markblaize) [inline contracts/dtoken/src/ft.rs:43]: @convowork1 user input should alway be validated

**Review by markblaize — CHANGES_REQUESTED** [review]

File: contracts/controller/src/liquidation.rs [hunk 6]
```diff
@@ -1,18 +1,94 @@
...
+        &mut self,
+        borrower: AccountId,
+        _borrowing_dtoken: AccountId,
+        collateral_dtoken: AccountId,
+        liquidator: AccountId,
+        liquidation_amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            self.increase_borrows(borrower.clone(), _borrowing_dtoken, liquidation_amount);
+            log!("Liquidation failed on borrow_repay call, revert changes...");
```
[the remark this answers is not in the document]
Comment (convowork1) [reply contracts/controller/src/liquidation.rs:68]: There is no this kind of logs in controller contract. It's only available for dtoken

**Review by markblaize — APPROVED** [review]

**Pull request merged**